lib/semantics/check-do-concurrent.cc

```diff
@@ -215,32 +216,91 @@ class FindDoConcurrentLoops {
+        DoConcurrentLabelEnforce doConcurrentLabelEnforce{
+            messages_, doConcurrentEnforcement.labels()};
+        parser::Walk(
+            std::get<parser::Block>(doConstruct.t), doConcurrentLabelEnforce);
+        EnforceConcurrentLoopControl(*concurrent);
+      } else if (auto *loopBounds{
+                     std::get_if<parser::LoopBounds<parser::ScalarIntExpr>>(
+                         &std::get<std::optional<parser::LoopControl>>(
+                             doStmt.statement.t)
+                              ->u)}) {
+        // C1120
+        auto *doVariable{loopBounds->name.thing.thing.symbol};
+        auto &source{loopBounds->name.thing.thing.source};
+        if (IntegerVariable(*doVariable)) {
+          // warning only: older Fortrans allowed floating-point do-variables
+          messages_.Say(source, "do-variable must have INTEGER type"_en_US);
+        }
+      } else {
+        // C1006
+        auto &logicalExpr{std::get<parser::ScalarLogicalExpr>(
+            std::get<std::optional<parser::LoopControl>>(doStmt.statement.t)->u)
+                              .thing.thing};
+        auto &source{doStmt.source};
+        (void)logicalExpr;  // is this expression common::TypeCategory::LOGICAL?
+        (void)source;
+      }
     }
-    return true;
   }
 
 private:
+  void EnforceConcurrentLoopControl(
+      const parser::LoopControl::Concurrent &concurrent) {
+    auto &header{std::get<parser::ConcurrentHeader>(concurrent.t)};
+    auto &mask{std::get<std::optional<parser::ScalarLogicalExpr>>(header.t)};
+    if (mask.has_value()) {
+      // C1121 - procedures in mask must be pure
+      // FIXME - implement
+    }
+    auto &controls{std::get<std::list<parser::ConcurrentControl>>(header.t)};
+    std::vector<semantics::Symbol> indexNames;
+    for (auto &c : controls) {
+      auto &name{std::get<parser::Name>(c.t)};
+      // C1122
+      if (!name.symbol) {
+        continue; // XXX - this shouldn't be needed
+      }
+      if (!IntegerVariable(*name.symbol)) {
+        messages_.Say(
+            name.source, "index-name must have INTEGER type"_err_en_US);
+      }
+      indexNames.push_back(*name.symbol);
+    }
+    if (!indexNames.empty()) {
+      for (auto &c : controls) {
+        // C1123
+        HasNoReferences(indexNames, std::get<1>(c.t));
+        HasNoReferences(indexNames, std::get<2>(c.t));
+        if (std::get<3>(c.t)) {
+          HasNoReferences(indexNames, *std::get<3>(c.t));
+        }
+      }
+    }
+  }
+  static void HasNoReferences(const std::vector<semantics::Symbol> &indexNames,
+      const parser::ScalarIntExpr &expression) {
+    // FIXME - implement
+  }
+  static bool IntegerVariable(const Symbol &variable) {
+    return variable.GetType()->category() !=
+        semantics::DeclTypeSpec::Intrinsic ||
+        variable.GetType()->intrinsicTypeSpec().category() !=
+        common::TypeCategory::Integer;
```

Comment:
It looks like this expression should be negated. Or maybe the function is meant to be `NotIntegerVariable`.